### ISSUE TITLE
Draw State layer does not remove a command buffer from the inFlightCmdBuffers set, even though it is implicitly waited upon with a vkWaitForFences() call, which later causes a crash

### ISSUE THREAD

**Issue by KarenGhavam** [KhronosGroup/Vulkan-LoaderAndValidationLayers#68]
This issue came from LunarXchange (#307). To be assigned to Google in github.

Description

I'm getting a crash at vkDeviceWaitIdle() call time in the Draw State layer code because of the layer making an invalid assumption that there are command buffers in flight, even though they have been released a long time ago.

The long-dead command buffers are the ones which generate the problem I described earlier today in #303.

What may be also important to note is that the situation, in which #303 arises, results from a single vkQueueSubmit() call which submits 2 command buffers. Recently, the function has been refactored, so that it now is possible to submit multiple command buffers with a single call, and the expectation is that the fence the user (optionally; in this particular case, I do pass a non-NULL handle) specifies is going to be set only after all the command buffers finish executing. Could it be that the debug call-back causes that particular command buffer to not be removed from the dev_data->inFlightCmdBuffers set?

Dialogue from LunarXchange:
Dean Sekulic

I've encoutered the same problem and right before I started to write report here it occured to me that I did something wrong; in this particular case, I had "long-dead" command buffer, but I forgot to wait for its fence before destroying it!

And then validation layer correctly(!) reported something in line that fence (of alive command buffer) was waiting for destroyed command buffer also.

HTH

Dominik Witczak

"Never say never" :) however, we do have an object tracker that keeps track of objects that have not gone out of scope yet, and it does not indicate the object could be dangling somewhere.

@Mark: should I quadruple-check this?

Dominik Witczak

Oh, and the app does wait on the fence :) Sorry, I missed that bit when I first read your message, Dean.

Tobine Ehlis

Dominik, can your run with GitHub master? If so, can you double-check and confirm if this is still an issue. I know some code changes related to this have gone in, but don't have a specific test case that was hitting your error.

**Comment by lentinem** [comment]
Would be great to have a repro here as the information crash in vkdevicewaitidle isn't really enough to repro this on our own. Also I assume this is still an issue?

**Comment by kbiElude** [comment]
Can't promise I'll be able to re-check this today. If I don't get back to you, I should be able to have a look on Weds (I'm OOO on Monday and Tues). Thanks.

**Comment by lentinem** [comment]
For some reason I can't assign to kbiElude; assigning to Karen so it doesn't get lost but we are waiting on an update for this one.

**Comment by KarenGhavam** [comment]
That's fine to leave it in my name for now. You can't assign it to Dominik because only members with write access to the repo can own issues.

**Comment by kbiElude** [comment]
I'm afraid I am unable to determine at this point, which app used to repro this problem. However, I managed to get a crash in VkLayer_draw_state. Fortunately, this time it should be trivial to patch:

![image](https://cloud.githubusercontent.com/assets/9154207/13664190/233aac04-e6a6-11e5-91d6-a3f813ef16d3.png)

**Comment by DominikWitczakAMD** [comment]
OK, I think I actually know where the originally reported error comes from! Consider the case described by #99. The layer reports the "attempt to free command buffer X which is in use" error for the *first* of the two command buffers I'm passing with a single vkQueueSubmit() call. It's as if the layer did not notice the command buffer had already finished executing, owing to the fact we did wait on the fence passed with the vkQueueSubmit() call, before proceeding with the cmd buffer destruction.

**Comment by lentinem** [comment]
The layer treats each element of the submit as independent so this should work properly. I'll take a look to make sure but if not we can fix the crash anyway.

**Comment by lentinem** [comment]
This was likely fixed by MR 110 but it would be good to verify.

**Comment by DominikWitczakAMD** [comment]
Works OK, safe to close.
